blog: Add PortSwigger lab walkthrough (Broken brute-force protection, IP block)

Wryhder opened the pull request:
@kingthorin I tried to add callouts/admonitions (called [alerts in the Hugo docs](https://gohugo.io/render-hooks/blockquotes/#alerts)) like this to match the look in the Notion doc, but it doesn't seem to work:

```
> [!NOTE]
> Useful information that users should know, even when skimming content.
``` 

Can I add a template for alerts as shown [in the docs here](https://gohugo.io/render-hooks/blockquotes/#example)?

psiinon commented:
![Logo](https://cdn.ast.checkmarx.net/integrations/logo/Checkmarx.png)
**Checkmarx One – Scan Summary & Details** – [7f2c4022\-6800\-4edc\-8e46\-e840ae8de27c](https://eu.ast.checkmarx.net/projects/f9121e40-c915-4b13-bb15-36d946943fd8/scans?id=7f2c4022-6800-4edc-8e46-e840ae8de27c&branch=Wryhder%2Fmain)
### Great job, no security vulnerabilities found in this Pull Request

kingthorin commented:
I'm fine with that being added.

@psiinon, @thc202 do we want those in a separate PR then rebase the blog on top after accepted/merged??

Wryhder commented:
Leaving a comment here about force-pushed changes. I added the sign-off trailer.

kingthorin commented:
Nope you're good, CLA isn't enabled for this repo. Thanks for checking though.

Wryhder commented:
@kingthorin I was working through a different lab and realized I got some details in this walkthrough mixed up with details from another lab. I'll correct the inaccuracies and let you know when this is ready for review again.

kingthorin commented:
Thanks for the heads up. I hadn't gone to look at the lab at all, so we easily could have missed that.....

kingthorin commented on site/content/blog/2025-02-17-portswigger-lab-walkthrough-broken-brute-force-protection-ip-block/index.md:
I guess you didn't test this?

![image](https://github.com/user-attachments/assets/c9c879eb-b245-4884-99dc-424f69bcb233)

Wryhder replied:
The callouts in the doc depends on changes added in #2977. I tested that PR with some sample content, but this PR doesn't contain those changes yet.

psiinon commented:
Thanks for this! As you've probably noticed we've got another Portswigger lab PR outstanding #2973 so we'll aim to get that one published first and then focus on this one 😁

kingthorin commented:
This can be rebased now

kingthorin commented on site/content/blog/2025-02-17-portswigger-lab-walkthrough-broken-brute-force-protection-ip-block/index.md:
Are these all really needed?

Wryhder replied:
Do you mean the "&nbsp;"? I used them to add some extra spacing where the text feels too tight against images or code blocks. I was thinking it'd be good for readability. Should I use CSS instead?

kingthorin replied:
Hmmm leave it for now, I’ll try it live tomorrow

kingthorin replied:
While I was going through this live/locally I removed them all and it seemed to render fine from my perspective.

kingthorin commented on site/content/blog/2025-02-17-portswigger-lab-walkthrough-broken-brute-force-protection-ip-block/index.md:
Seems to be some extra leading space

kingthorin commented on site/content/blog/2025-02-17-portswigger-lab-walkthrough-broken-brute-force-protection-ip-block/index.md:
The hyphen notes at the end don't really seem to add much

kingthorin reviewed (commented):
I'll try to do a thorough review of the text and details over the next few days.

Wryhder commented:
> I'll try to do a thorough review of the text and details over the next few days.

Alright, thank you!

Wryhder commented on site/content/blog/2025-02-17-portswigger-lab-walkthrough-broken-brute-force-protection-ip-block/index.md:
@kingthorin Just marking out sections where I'd especially appreciate your help whenever you can take a look. I tried to make sure the ZAP-specific details are accurate, but I'm not too certain about some parts.

Wryhder commented on site/content/blog/2025-02-17-portswigger-lab-walkthrough-broken-brute-force-protection-ip-block/index.md:
Here, too. Thanks a bunch!

kingthorin replied:
This is fine. On macOS we do bundle a JRE, but on other platforms we don't so this statement is fine in my opinion.

kingthorin commented on site/content/blog/2025-02-17-portswigger-lab-walkthrough-broken-brute-force-protection-ip-block/index.md:
Confirmed: I removed your a tag and ran locally, `http://localhost:3000/blog/2025-02-17-portswigger-lab-walkthrough-broken-brute-force-protection-ip-block/#have-username-got-password` work just fine.

kingthorin commented on site/content/blog/2025-02-17-portswigger-lab-walkthrough-broken-brute-force-protection-ip-block/index.md:
```diff
@@ -0,0 +1,793 @@
+title: "PortSwigger Lab Walkthrough: Broken Brute-Force Protection, IP Block"
+summary: >
+  Walkthrough for the PortSwigger lab, "Broken brute-force protection, IP block".
+# images:
+# - 
```
It would be great if you could some up with at least one so that it can be used for the social preview

kingthorin commented on site/content/blog/2025-02-17-portswigger-lab-walkthrough-broken-brute-force-protection-ip-block/index.md:
"...brute-forcing carlos' password."

kingthorin replied:
I've suggested "carlos'" which you seem to have used quite a few places as well. However there's also some "carlos's". I'm not terribly set either way, but I'd suggest being consistent so do a search and replace one way or the other :wink:

kingthorin commented on site/content/blog/2025-02-17-portswigger-lab-walkthrough-broken-brute-force-protection-ip-block/index.md:
I'd suggest making the second one here an inline code block instead of a link, since the link is yours, and likely not available any longer or to anyone else:
`https://0a19000b0319d4b784b7a40400f800ed.web-security-academy.net/`
(Inline code uses single backticks before and after)

kingthorin commented on site/content/blog/2025-02-17-portswigger-lab-walkthrough-broken-brute-force-protection-ip-block/index.md:
"...were used over the course of writing this article:"

kingthorin commented on site/content/blog/2025-02-17-portswigger-lab-walkthrough-broken-brute-force-protection-ip-block/index.md:
I'd suggest "ZAP" vs "Zaproxy" for the hyphen notes..... though maybe those aren't necessary here either? 🤷‍♂️

kingthorin commented on site/content/blog/2025-02-17-portswigger-lab-walkthrough-broken-brute-force-protection-ip-block/index.md:
I'd suggest a single backtick inline code here instead of quoting it.

kingthorin commented on site/content/blog/2025-02-17-portswigger-lab-walkthrough-broken-brute-force-protection-ip-block/index.md:
Double space after "can"

kingthorin commented on site/content/blog/2025-02-17-portswigger-lab-walkthrough-broken-brute-force-protection-ip-block/index.md:
Two spaces before "enables us"

Wryhder replied:
@kingthorin RE:

> Yeah, but then we also need to login to our own account (wiener:peter) every third login attempt.

kingthorin commented on site/content/blog/2025-02-17-portswigger-lab-walkthrough-broken-brute-force-protection-ip-block/index.md:
I thought you said earlier we knew for sure that it was "carlos" and we only needed to sort out the password?

Wryhder replied:
Yeah, but then we also need to login to our own account (`wiener:peter`) every third login attempt.

kingthorin commented on site/content/blog/2025-02-17-portswigger-lab-walkthrough-broken-brute-force-protection-ip-block/index.md:
If you've generated the lists there's no need to use a script for the payloads, just use the text files.

kingthorin commented on site/content/blog/2025-02-17-portswigger-lab-walkthrough-broken-brute-force-protection-ip-block/index.md:
Space before the period

kingthorin commented on site/content/blog/2025-02-17-portswigger-lab-walkthrough-broken-brute-force-protection-ip-block/index.md:
Now I see why you went script not files, maybe this should be clarified earlier.

kingthorin replied:
BTW: Thanks for linking my SO answer 😁

kingthorin commented on site/content/blog/2025-02-17-portswigger-lab-walkthrough-broken-brute-force-protection-ip-block/index.md:
Which is it? The text and screenshots need to be updated with the correct finding.

Wryhder replied:
Ah, I wasn't sure how to handle this section. Since I solved the lab twice with two different methods, there are two different passwords and slightly different ways to locate them (since the first password is mixed up in other 302 results, unlike the second). The images show this difference and the text is suppose to complement them. Do you think I should approach this differently?

kingthorin replied:
Do both passwords actually solve the lab?

Wryhder replied:
Yeah, both solutions are valid. But the actual password is not necessarily going to be the same for each reader. I've solved some labs multiple times. Same method, different lab instances, different passwords. (I think each lab instance is available for a couple hours max, so if you're done by then, you have to start up a new one). Not sure about the technical details of how the labs are set up.

kingthorin replied:
Ok so just make sure to add a callout note/warning that clarifies this.

Wryhder replied:
Makes sense, thanks.

kingthorin commented on site/content/blog/2025-02-17-portswigger-lab-walkthrough-broken-brute-force-protection-ip-block/index.md:
ZAP Docs

kingthorin commented:
To address the DCO requirement you'll need to sign-off the commit(s):

- https://github.com/zaproxy/zaproxy/blob/main/CONTRIBUTING.md#developer-certificate-of-origin
- https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff

Wryhder commented:
@kingthorin I've read through your comments (left replies on some). I'll address them within the next few days. Thanks for being so thorough with your review, Appreciate it!

Wryhder commented:
@kingthorin I've implemented all feedback from your review. Also fixed the DCO issue (I had the trailer in the previous commits but was modifying it manually and it ended up being invalid).

Wryhder commented on site/content/blog/2025-02-17-portswigger-lab-walkthrough-broken-brute-force-protection-ip-block/index.md:
```diff
@@ -0,0 +1,755 @@
+summary: >
+  Walkthrough for the PortSwigger lab, "Broken brute-force protection, IP block".
+images:
+- https://www.zaproxy.org/blog/2025-02-17-portswigger-lab-walkthrough-broken-brute-force-protection-ip-block/images/image header-key and house models.jpg
```
@kingthorin I followed the example of the other articles here. Is this correct?

kingthorin replied:
Yeah, I'd suggest using underscores or hyphens in the filename, and reducing the length (it seems overly verbose).
(Here and the file itself obviously :wink:)

Wryhder replied:
> reducing the length (it seems overly verbose)

Just to be clear, you mean just the filename of the image, yes? Not the entire article (that one's long, too)?

kingthorin replied:
Ya just the image

Wryhder commented on site/content/blog/2025-02-17-portswigger-lab-walkthrough-broken-brute-force-protection-ip-block/index.md:
I needed to include some attribution but wasn't sure where to add it. Is this okay?

Oh, and what do you think of the image itself?

kingthorin replied:
I'm good with this, and I'm good with the image too.

thc202 replied:
The license says that attribution is not required:
https://www.pexels.com/license/

> Attribution is not required. Giving credit to the photographer or Pexels is not necessary but always appreciated.

Not that I don't mind the attribution just saying.

kingthorin commented:
Assuming that one filename/reference thing is addressed I'm good with this article now.

Wryhder commented:
@kingthorin I've made the filename change. Also squashed a few commits.

Wryhder commented:
Forced-pushed change is more squashing. Done now.

kingthorin reviewed (approved):
Thanks!

Wryhder commented:
@kingthorin My apologies. Missed a couple of spots.

psiinon commented:
I've updated the date to today and made some minor tweaks to address @thc202's feedback

thc202 commented:
Thank you both!

psiinon commented:
Thanks @Wryhder and sorry this took so long to get published.
Its now live on https://www.zaproxy.org/blog/2025-04-09-portswigger-labs-broken-brute-force-protection-ip-block/ and shared via the usual social media sites 😁

Wryhder commented:
@thc202 Thanks a bunch for your review.

Thanks @psiinon for making the changes. Article looks good, thank you!

And thanks again @kingthorin for your help with this.
